Source/WebCore/workers/service/background-fetch/ServiceWorkerRegistrationBackgroundFetchAPI.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ServiceWorkerRegistrationBackgroundFetchAPI.h"
+
+#if ENABLE(SERVICE_WORKER)
+
+#include "BackgroundFetchManager.h"
+#include "ServiceWorkerRegistration.h"
+#include <wtf/StdLibExtras.h>
+
+namespace WebCore {
+
+ServiceWorkerRegistrationBackgroundFetchAPI::ServiceWorkerRegistrationBackgroundFetchAPI(ServiceWorkerRegistration& serviceWorkerRegistration)
+    : m_serviceWorkerRegistration(serviceWorkerRegistration)
+{
+}
+
+ServiceWorkerRegistrationBackgroundFetchAPI::~ServiceWorkerRegistrationBackgroundFetchAPI()
+{
+}
+
+BackgroundFetchManager& ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetch(ServiceWorkerRegistration& serviceWorkerRegistration)
+{
+    return ServiceWorkerRegistrationBackgroundFetchAPI::from(serviceWorkerRegistration)->backgroundFetchManager();
+}
+
+RefPtr<BackgroundFetchManager> ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetchIfCreated(ServiceWorkerRegistration& serviceWorkerRegistration)
+{
+    return ServiceWorkerRegistrationBackgroundFetchAPI::from(serviceWorkerRegistration)->m_backgroundFetchManager.get();
+}
+
+BackgroundFetchManager& ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetchManager()
+{
+    if (!m_backgroundFetchManager)
+        m_backgroundFetchManager = BackgroundFetchManager::create(m_serviceWorkerRegistration);
+
+    return *m_backgroundFetchManager;
+}
+
+ServiceWorkerRegistrationBackgroundFetchAPI* ServiceWorkerRegistrationBackgroundFetchAPI::from(ServiceWorkerRegistration& serviceWorkerRegistration)
```

Comment:
Should probably return a reference instead of a raw pointer.

Reply:
right